# Session 9

## 7 Other Group of Functions in pandas

### 1- Importing Data



Reading Path 3 ways

In [ ]:
path = 'movies.csv'  # csv  comma seperated values
data = pd.read_csv(path)       #mahmoud,25,1990  mahmoud 25 1990
# data

In [ ]:
path = 'D:\Temp/movies.csv'  # Full path csv  comma seperated values
data = pd.read_csv(path)       #mahmoud,25,1990  mahmoud 25 1990
# data

In [ ]:
path = '../data/movies.csv'  # Full path csv  comma seperated values
data = pd.read_csv(path)       #mahmoud,25,1990  mahmoud 25 1990
# data

In [ ]:
data

title     rating  year  runtime
0             Almost Famous          R  2000    122.0
1              American Pie          R  1999     95.0
2        Back to the Future         PG  1985    116.0
3              Blade Runner          R  1982    117.0
4         Blood for Dracula          R  1974    106.0
5               Blue Velvet          R  1986    120.0
6        The Breakfast Club          R  1985     97.0
7                  Clueless      PG-13  1995      NaN
8            Cool Hand Luke         GP  1967    127.0
9                 The Craft          R  1996    101.0
10           Doctor Zhivago      PG-13  1965    197.0
11                  el Topo  Not Rated  1970    125.0
12                Evil Dead      NC-17  1981     85.0
13             Ghostbusters         PG  1984    105.0
14                   Grease      PG-13  1978    110.0
15                 Heathers          R  1988    103.0
16                Labyrinth         PG  1986    101.0
17            The Lost Boys          R  1987     97.0
18               Mean Girls      PG-13  2004     97.0
19       Millennium Actress         PG  2001     87.0
20       My Neighbor Totoro          G  1988     86.0
21        Napoleon Dynamite         PG  2004     96.0
22  Pee-wee's Big Adventure         PG  1985     91.0
23           Pretty in Pink      PG-13  1986     97.0
24       The Princess Bride         PG  1987     98.0
25                   Psycho          R  1960      NaN
26              Stand by Me          R  1986     89.0
27                  Super 8      PG-13  2011    112.0
28                 superbad          R  2007    113.0
29                 WarGames         PG  1983    114.0

In [ ]:
# pd.read_csv(filename)	From a CSV file

In [ ]:
# pd.read_excel(filename)	From an Excel file    xlsx format

In [ ]:
# pd.read_sql(query, connection_object)	Read from a SQL table/database

### 2- Viewing/exploring Data

In [ ]:
import pandas as pd

In [ ]:
path = 'D:\Work\Cources\AI Academy\G4 Summer offline\Session 6/movies.csv'  # Full path csv  comma seperated values
data = pd.read_csv(path)       #mahmoud,25,1990  mahmoud 25 1990
# data

In [ ]:
data

title     rating  year  runtime
0             Almost Famous          R  2000    122.0
1              American Pie          R  1999     95.0
2        Back to the Future         PG  1985    116.0
3              Blade Runner          R  1982    117.0
4         Blood for Dracula          R  1974    106.0
5               Blue Velvet          R  1986    120.0
6        The Breakfast Club          R  1985     97.0
7                  Clueless      PG-13  1995      NaN
8            Cool Hand Luke         GP  1967    127.0
9                 The Craft          R  1996    101.0
10           Doctor Zhivago      PG-13  1965    197.0
11                  el Topo  Not Rated  1970    125.0
12                Evil Dead      NC-17  1981     85.0
13             Ghostbusters         PG  1984    105.0
14                   Grease      PG-13  1978    110.0
15                 Heathers          R  1988    103.0
16                Labyrinth         PG  1986    101.0
17            The Lost Boys          R  1987     97.0
18               Mean Girls      PG-13  2004     97.0
19       Millennium Actress         PG  2001     87.0
20       My Neighbor Totoro          G  1988     86.0
21        Napoleon Dynamite         PG  2004     96.0
22  Pee-wee's Big Adventure         PG  1985     91.0
23           Pretty in Pink      PG-13  1986     97.0
24       The Princess Bride         PG  1987     98.0
25                   Psycho          R  1960      NaN
26              Stand by Me          R  1986     89.0
27                  Super 8      PG-13  2011    112.0
28                 superbad          R  2007    113.0
29                 WarGames         PG  1983    114.0

In [ ]:
df = data

In [ ]:
df = data.copy()   # copy because not to overwrite orignial data

 Solution

create a deep copy of the DataFrame. This ensures that any modifications made to the new DataFrame (df) do not affect the original DataFrame (data).

Here’s a detailed explanation:

- Avoid Unintended Modifications: When you assign df = data, both df and data refer to the same DataFrame in memory. Any changes made to df will also affect data and vice versa. By using df = data.copy(), you create a separate copy of the DataFrame. This allows you to modify df without altering data.

- Preserve Original Data: Keeping the original data unchanged is crucial, especially when you need to compare the modified DataFrame with the original or when the original DataFrame is required for further operations.

- Ensure Data Integrity: In data analysis and preprocessing, it’s common to perform various transformations and experiments on the data. Creating a copy helps ensure that you can always revert to the original dataset if needed.

In [ ]:
# df = data     # Forbidden 

In [ ]:
df.head() #First n rows of the DataFrame

title rating  year  runtime
0       Almost Famous      R  2000    122.0
1        American Pie      R  1999     95.0
2  Back to the Future     PG  1985    116.0
3        Blade Runner      R  1982    117.0
4   Blood for Dracula      R  1974    106.0

In [ ]:
df.tail() #Last n rows of the DataFrame

title rating  year  runtime
25       Psycho      R  1960      NaN
26  Stand by Me      R  1986     89.0
27      Super 8  PG-13  2011    112.0
28     superbad      R  2007    113.0
29     WarGames     PG  1983    114.0

In [ ]:
df.sample(5)

title rating  year  runtime
12      Evil Dead  NC-17  1981     85.0
9       The Craft      R  1996    101.0
14         Grease  PG-13  1978    110.0
13   Ghostbusters     PG  1984    105.0
0   Almost Famous      R  2000    122.0

In [ ]:
df.shape   #Number of rows and columns

(30, 4)

In [ ]:
df.info() #Index, Datatype and Memory information

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 30 entries, 0 to 29
Data columns (total 4 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   title    30 non-null     object 
 1   rating   30 non-null     object 
 2   year     30 non-null     int64  
 3   runtime  28 non-null     float64
dtypes: float64(1), int64(1), object(2)
memory usage: 1.1+ KB


**Insights** 

- 1- data has 30 rows and 4 columns

- 2- range index (0 to 29) means there is no missing data

- 3- column runtime has 2 nulls 

- 4- data has 2 object data type

In [ ]:
df['rating'].value_counts() # apply only on categorical (string) features 

rating
R            12
PG            8
PG-13         6
GP            1
Not Rated     1
NC-17         1
G             1
Name: count, dtype: int64

Task

how to get percentage of counting for every value

In [ ]:
df['rating'].value_counts(normalize =True)

rating
R            0.400000
PG           0.266667
PG-13        0.200000
GP           0.033333
Not Rated    0.033333
NC-17        0.033333
G            0.033333
Name: proportion, dtype: float64

R rating is the most repeated 

task

1- read dataframe

2- display information

3- print last 5 rows

4- print shape of it

5- write info of data

### 3- Statistics

In [ ]:
df.describe(include='all')   #Summary statistics for all columns 

title rating         year     runtime
count              30     30    30.000000   28.000000
unique             30      7          NaN         NaN
top     Almost Famous      R          NaN         NaN
freq                1     12          NaN         NaN
mean              NaN    NaN  1987.000000  107.285714
std               NaN    NaN    12.406895   21.308766
min               NaN    NaN  1960.000000   85.000000
25%               NaN    NaN  1982.250000   96.750000
50%               NaN    NaN  1986.000000  102.000000
75%               NaN    NaN  1995.750000  114.500000
max               NaN    NaN  2011.000000  197.000000

**Insights**

- 1- title feature has 30 values with 30 unique values
- 2- Rating Feature has 30 values with 7 unique values
- 3- Rating Feature has the most freq value is R with 12 times
- 4- min and max of year feature is 1960 and 2011 respectively ......
- 5- runtime feature has 28 values with 2 missing nulls.
- 6- the average of time of movies is almost 107 mintues
- 7- the shortest movie is 85 minute with it is name  .....
- 8- the longest movie with 197 mintue is ..... 

Task  : 

1- find the shortest and longest movie name 

2- which movie is the oldest and newest

In [ ]:
df['runtime'].min()

85.0

In [ ]:
df.loc[ df['runtime'] == df['runtime'].min() ,'title']

12    Evil Dead
Name: title, dtype: object

In [ ]:
df[ df['runtime'] == df['runtime'].min() ]['title']

12    Evil Dead
Name: title, dtype: object

In [ ]:
df[ df['runtime'] == df['runtime'].min() ]['title'].values[0]

'Evil Dead'

In [ ]:
# df['title']

In [ ]:
df[df['runtime'] == df['runtime'].min() ]['title']

12    Evil Dead
Name: title, dtype: object

In [ ]:
df[df['runtime'] == df['runtime'].max() ]['title']

10    Doctor Zhivago
Name: title, dtype: object

In [ ]:
df[['title','runtime']]

title  runtime
0             Almost Famous    122.0
1              American Pie     95.0
2        Back to the Future    116.0
3              Blade Runner    117.0
4         Blood for Dracula    106.0
5               Blue Velvet    120.0
6        The Breakfast Club     97.0
7                  Clueless      NaN
8            Cool Hand Luke    127.0
9                 The Craft    101.0
10           Doctor Zhivago    197.0
11                  el Topo    125.0
12                Evil Dead     85.0
13             Ghostbusters    105.0
14                   Grease    110.0
15                 Heathers    103.0
16                Labyrinth    101.0
17            The Lost Boys     97.0
18               Mean Girls     97.0
19       Millennium Actress     87.0
20       My Neighbor Totoro     86.0
21        Napoleon Dynamite     96.0
22  Pee-wee's Big Adventure     91.0
23           Pretty in Pink     97.0
24       The Princess Bride     98.0
25                   Psycho      NaN
26              Stand by Me     89.0
27                  Super 8    112.0
28                 superbad    113.0
29                 WarGames    114.0

In [ ]:
df[df['runtime'] == df['runtime'].max() ][['title','runtime']]

title  runtime
10  Doctor Zhivago    197.0

In [ ]:
df[df['year'] == df['year'].min() ][['title','year']]

title  year
25  Psycho  1960

In [ ]:
df[df['year'] == df['year'].max() ][['title','year']]

title  year
27  Super 8  2011

In [ ]:
df['runtime'].mean()	#Returns the mean of all columns

107.28571428571429

In [ ]:
df['runtime'].count()	#Returns the number of non-null values in each DataFrame column

28

In [ ]:
df['runtime'].max()	#Returns the highest value in each column

197.0

In [ ]:
df['runtime'].min()	#Returns the lowest value in each column

85.0

In [ ]:
df['runtime'].median()	#Returns the median of each column

102.0

In [ ]:
df['runtime'].std()	#Returns the standard deviation of each column

21.308765615668406

In [ ]:
df

title     rating  year  runtime
0             Almost Famous          R  2000    122.0
1              American Pie          R  1999     95.0
2        Back to the Future         PG  1985    116.0
3              Blade Runner          R  1982    117.0
4         Blood for Dracula          R  1974    106.0
5               Blue Velvet          R  1986    120.0
6        The Breakfast Club          R  1985     97.0
7                  Clueless      PG-13  1995      NaN
8            Cool Hand Luke         GP  1967    127.0
9                 The Craft          R  1996    101.0
10           Doctor Zhivago      PG-13  1965    197.0
11                  el Topo  Not Rated  1970    125.0
12                Evil Dead      NC-17  1981     85.0
13             Ghostbusters         PG  1984    105.0
14                   Grease      PG-13  1978    110.0
15                 Heathers          R  1988    103.0
16                Labyrinth         PG  1986    101.0
17            The Lost Boys          R  1987     97.0
18               Mean Girls      PG-13  2004     97.0
19       Millennium Actress         PG  2001     87.0
20       My Neighbor Totoro          G  1988     86.0
21        Napoleon Dynamite         PG  2004     96.0
22  Pee-wee's Big Adventure         PG  1985     91.0
23           Pretty in Pink      PG-13  1986     97.0
24       The Princess Bride         PG  1987     98.0
25                   Psycho          R  1960      NaN
26              Stand by Me          R  1986     89.0
27                  Super 8      PG-13  2011    112.0
28                 superbad          R  2007    113.0
29                 WarGames         PG  1983    114.0

task Data anaylsis 


- average of  movies with rating R

- calcualte the average of movies period that published before 1987

Bonus

- get movies  that runtime <90 or rating = R

- calcualte the max of movies period that published after 1995 and rating R

In [ ]:
df[df['rating'] == 'R']['runtime'].mean()

105.45454545454545

In [ ]:
df[df['year'] < 1987]['runtime'].mean()

112.3125

In [ ]:
df[(df['year'] > 1995) & (df['rating'] == 'R')]['runtime'].max()

122.0

In [ ]:
df[ (df['runtime'] < 90)| (df['rating'] == 'R') ]['title'].values

array(['Almost Famous', 'American Pie', 'Blade Runner',
       'Blood for Dracula', 'Blue Velvet', 'The Breakfast Club',
       'The Craft', 'Evil Dead', 'Heathers', 'The Lost Boys',
       'Millennium Actress', 'My Neighbor Totoro', 'Psycho',
       'Stand by Me', 'superbad'], dtype=object)

In [ ]:
df.corr(numeric_only=True)	#Returns the correlation between columns in a DataFrame

year   runtime
year     1.000000 -0.450278
runtime -0.450278  1.000000

### 4- Data Cleaning



In [ ]:
df.sample(5)

title rating  year  runtime
14          Grease  PG-13  1978    110.0
23  Pretty in Pink  PG-13  1986     97.0
18      Mean Girls  PG-13  2004     97.0
0    Almost Famous      R  2000    122.0
10  Doctor Zhivago  PG-13  1965    197.0

In [ ]:
df.columns

Index(['title', 'rating', 'year', 'runtime'], dtype='object')

In [ ]:
df.columns = ['Title','Rating','Year','Runtime']  #Rename columns
df.sample(2)

Title Rating  Year  Runtime
12  Evil Dead  NC-17  1981     85.0
9   The Craft      R  1996    101.0

In [ ]:
df.rename(columns={'Year':'YearProd'},inplace=True)

In [ ]:
df.sample()

Title Rating  YearProd  Runtime
18  Mean Girls  PG-13      2004     97.0

task

write list comphrension to edit column names + '_'

In [ ]:
df.columns

Index(['Title', 'Rating', 'YearProd', 'Runtime'], dtype='object')

In [ ]:
df.columns = [  col + '_'   for col in df.columns ]        
              # operation        for loop           condition

In [ ]:
df.columns = [  col.upper()    for col in df.columns ]        
              # operation        for loop           condition

In [ ]:
# 'Run time'.replace(' ','_')

In [ ]:
df.columns

Index(['TITLE_', 'RATING_', 'YEARPROD_', 'RUNTIME_'], dtype='object')

In [ ]:
df.sample()

TITLE_ RATING_  YEARPROD_  RUNTIME_
24  The Princess Bride      PG       1987      98.0

In [ ]:
# df.columns = [  col.upper()  for col in df.columns ]  
# df.sample()

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 30 entries, 0 to 29
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   TITLE_     30 non-null     object 
 1   RATING_    30 non-null     object 
 2   YEARPROD_  30 non-null     int64  
 3   RUNTIME_   28 non-null     float64
dtypes: float64(1), int64(1), object(2)
memory usage: 1.1+ KB


In [ ]:
df['TITLE_'] = df['TITLE_'].astype('string') # str	#Convert the datatype of the series to float

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 30 entries, 0 to 29
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   TITLE_     30 non-null     string 
 1   RATING_    30 non-null     object 
 2   YEARPROD_  30 non-null     int64  
 3   RUNTIME_   28 non-null     float64
dtypes: float64(1), int64(1), object(1), string(1)
memory usage: 1.1+ KB


In [ ]:
df['RATING_'] = df['RATING_'].astype('category') # str	#Convert the datatype of the series to float
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 30 entries, 0 to 29
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype   
---  ------     --------------  -----   
 0   TITLE_     30 non-null     string  
 1   RATING_    30 non-null     category
 2   YEARPROD_  30 non-null     int64   
 3   RUNTIME_   28 non-null     float64 
dtypes: category(1), float64(1), int64(1), string(1)
memory usage: 1.2 KB


# Task



1- Read ecommerce data

2- print info 

3- convert any column name to lower case and without any space

4- what is the average age of new customers 

5- what is total sales of gender female



In [ ]:
eccom = 'D:\Work\Cources\AI Academy\G4 Summer offline\Session 6/ecommerce_sales.csv'
ecom = pd.read_csv(eccom)


In [ ]:
ecom.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 10 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Company_ID        10000 non-null  int64  
 1   Purchase_Date     10000 non-null  object 
 2   Product_Category  10000 non-null  object 
 3   Quantity          10000 non-null  int64  
 4   Total_Sales       10000 non-null  float64
 5   Gender            10000 non-null  object 
 6   Marital_Status    10000 non-null  object 
 7   Price_Per_Unit    10000 non-null  float64
 8   Customer_Type     10000 non-null  object 
 9   Age               10000 non-null  int64  
dtypes: float64(2), int64(3), object(5)
memory usage: 781.4+ KB


In [ ]:
ecom.columns = [ col.lower().replace(' ','_')   for col in ecom.columns]
ecom.sample()

company_id purchase_date product_category  quantity  total_sales gender  \
7439           4    2020-01-05           Beauty         8   362.350851   Male   

     marital_status  price_per_unit customer_type  age  
7439        Widowed       28.223462  New Customer   30

In [ ]:
ecom[ecom['customer_type'] == 'New Customer' ]['age'].mean()

48.31803734604688

In [ ]:
ecom[ecom['gender'] == 'Female' ]['total_sales'].sum()

1282081.6326928153

In [ ]:
import pandas as pd

In [ ]:
path = 'D:\Work\Cources\AI Academy\G4 Summer offline\Session 6/movies.csv'  # Full path csv  comma seperated values
df = pd.read_csv(path)  

In [ ]:
df.nunique() # number of uniques 

title      30
rating      7
year       22
runtime    24
dtype: int64

In [ ]:
df['YEAR_'] = df['YEAR_'].astype(float)	#Convert the datatype of the series to float

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 30 entries, 0 to 29
Data columns (total 4 columns):
 #   Column    Non-Null Count  Dtype   
---  ------    --------------  -----   
 0   Title_    30 non-null     string  
 1   Rating_   30 non-null     category
 2   Year_     30 non-null     int64   
 3   Runtime_  28 non-null     float64 
dtypes: category(1), float64(1), int64(1), string(1)
memory usage: 1.2 KB


In [ ]:
df

title     rating  year  runtime
0             Almost Famous          R  2000    122.0
1              American Pie          R  1999     95.0
2        Back to the Future         PG  1985    116.0
3              Blade Runner          R  1982    117.0
4         Blood for Dracula          R  1974    106.0
5               Blue Velvet          R  1986    120.0
6        The Breakfast Club          R  1985     97.0
7                  Clueless      PG-13  1995      NaN
8            Cool Hand Luke         GP  1967    127.0
9                 The Craft          R  1996    101.0
10           Doctor Zhivago      PG-13  1965    197.0
11                  el Topo  Not Rated  1970    125.0
12                Evil Dead      NC-17  1981     85.0
13             Ghostbusters         PG  1984    105.0
14                   Grease      PG-13  1978    110.0
15                 Heathers          R  1988    103.0
16                Labyrinth         PG  1986    101.0
17            The Lost Boys          R  1987     97.0
18               Mean Girls      PG-13  2004     97.0
19       Millennium Actress         PG  2001     87.0
20       My Neighbor Totoro          G  1988     86.0
21        Napoleon Dynamite         PG  2004     96.0
22  Pee-wee's Big Adventure         PG  1985     91.0
23           Pretty in Pink      PG-13  1986     97.0
24       The Princess Bride         PG  1987     98.0
25                   Psycho          R  1960      NaN
26              Stand by Me          R  1986     89.0
27                  Super 8      PG-13  2011    112.0
28                 superbad          R  2007    113.0
29                 WarGames         PG  1983    114.0

try to replace 8   to  eight

In [ ]:
df['title'].replace('8','Eight',inplace=False)	#Replace all values equal to 8 with 'Eight'

0               Almost Famous
1                American Pie
2          Back to the Future
3                Blade Runner
4           Blood for Dracula
5                 Blue Velvet
6          The Breakfast Club
7                    Clueless
8              Cool Hand Luke
9                   The Craft
10             Doctor Zhivago
11                    el Topo
12                  Evil Dead
13               Ghostbusters
14                     Grease
15                   Heathers
16                  Labyrinth
17              The Lost Boys
18                 Mean Girls
19         Millennium Actress
20         My Neighbor Totoro
21          Napoleon Dynamite
22    Pee-wee's Big Adventure
23             Pretty in Pink
24         The Princess Bride
25                     Psycho
26                Stand by Me
27                    Super 8
28                   superbad
29                   WarGames
Name: title, dtype: object

In [ ]:
df['title'].replace('8','Eight',
           regex=True,      # regular exp.
           inplace=False)	#Replace all values equal to 8 with 'Eight'

0               Almost Famous
1                American Pie
2          Back to the Future
3                Blade Runner
4           Blood for Dracula
5                 Blue Velvet
6          The Breakfast Club
7                    Clueless
8              Cool Hand Luke
9                   The Craft
10             Doctor Zhivago
11                    el Topo
12                  Evil Dead
13               Ghostbusters
14                     Grease
15                   Heathers
16                  Labyrinth
17              The Lost Boys
18                 Mean Girls
19         Millennium Actress
20         My Neighbor Totoro
21          Napoleon Dynamite
22    Pee-wee's Big Adventure
23             Pretty in Pink
24         The Princess Bride
25                     Psycho
26                Stand by Me
27                Super Eight
28                   superbad
29                   WarGames
Name: title, dtype: object

In [ ]:
df['year'].replace(1960,1959,inplace=False)	#Replace all values equal to 1960 with 1959

0     2000
1     1999
2     1985
3     1982
4     1974
5     1986
6     1985
7     1995
8     1967
9     1996
10    1965
11    1970
12    1981
13    1984
14    1978
15    1988
16    1986
17    1987
18    2004
19    2001
20    1988
21    2004
22    1985
23    1986
24    1987
25    1959
26    1986
27    2011
28    2007
29    1983
Name: year, dtype: int64

In [ ]:
import numpy as np

In [ ]:
df['runtime'].replace(np.nan,0,inplace=False)	#Replace all values equal to 8 with 'Eight'

title     rating  year  runtime
0             Almost Famous          R  2000    122.0
1              American Pie          R  1999     95.0
2        Back to the Future         PG  1985    116.0
3              Blade Runner          R  1982    117.0
4         Blood for Dracula          R  1974    106.0
5               Blue Velvet          R  1986    120.0
6        The Breakfast Club          R  1985     97.0
7                  Clueless      PG-13  1995      0.0
8            Cool Hand Luke         GP  1967    127.0
9                 The Craft          R  1996    101.0
10           Doctor Zhivago      PG-13  1965    197.0
11                  el Topo  Not Rated  1970    125.0
12                Evil Dead      NC-17  1981     85.0
13             Ghostbusters         PG  1984    105.0
14                   Grease      PG-13  1978    110.0
15                 Heathers          R  1988    103.0
16                Labyrinth         PG  1986    101.0
17            The Lost Boys          R  1987     97.0
18               Mean Girls      PG-13  2004     97.0
19       Millennium Actress         PG  2001     87.0
20       My Neighbor Totoro          G  1988     86.0
21        Napoleon Dynamite         PG  2004     96.0
22  Pee-wee's Big Adventure         PG  1985     91.0
23           Pretty in Pink      PG-13  1986     97.0
24       The Princess Bride         PG  1987     98.0
25                   Psycho          R  1960      0.0
26              Stand by Me          R  1986     89.0
27                  Super 8      PG-13  2011    112.0
28                 superbad          R  2007    113.0
29                 WarGames         PG  1983    114.0

In [ ]:
df.set_index('year',inplace=False)	#Change the index

title     rating  runtime
year                                             
2000            Almost Famous          R    122.0
1999             American Pie          R     95.0
1985       Back to the Future         PG    116.0
1982             Blade Runner          R    117.0
1974        Blood for Dracula          R    106.0
1986              Blue Velvet          R    120.0
1985       The Breakfast Club          R     97.0
1995                 Clueless      PG-13      NaN
1967           Cool Hand Luke         GP    127.0
1996                The Craft          R    101.0
1965           Doctor Zhivago      PG-13    197.0
1970                  el Topo  Not Rated    125.0
1981                Evil Dead      NC-17     85.0
1984             Ghostbusters         PG    105.0
1978                   Grease      PG-13    110.0
1988                 Heathers          R    103.0
1986                Labyrinth         PG    101.0
1987            The Lost Boys          R     97.0
2004               Mean Girls      PG-13     97.0
2001       Millennium Actress         PG     87.0
1988       My Neighbor Totoro          G     86.0
2004        Napoleon Dynamite         PG     96.0
1985  Pee-wee's Big Adventure         PG     91.0
1986           Pretty in Pink      PG-13     97.0
1987       The Princess Bride         PG     98.0
1960                   Psycho          R      NaN
1986              Stand by Me          R     89.0
2011                  Super 8      PG-13    112.0
2007                 superbad          R    113.0
1983                 WarGames         PG    114.0

In [ ]:
df

title     rating  year  runtime
0             Almost Famous          R  2000    122.0
1              American Pie          R  1999     95.0
2        Back to the Future         PG  1985    116.0
3              Blade Runner          R  1982    117.0
4         Blood for Dracula          R  1974    106.0
5               Blue Velvet          R  1986    120.0
6        The Breakfast Club          R  1985     97.0
7                  Clueless      PG-13  1995      NaN
8            Cool Hand Luke         GP  1967    127.0
9                 The Craft          R  1996    101.0
10           Doctor Zhivago      PG-13  1965    197.0
11                  el Topo  Not Rated  1970    125.0
12                Evil Dead      NC-17  1981     85.0
13             Ghostbusters         PG  1984    105.0
14                   Grease      PG-13  1978    110.0
15                 Heathers          R  1988    103.0
16                Labyrinth         PG  1986    101.0
17            The Lost Boys          R  1987     97.0
18               Mean Girls      PG-13  2004     97.0
19       Millennium Actress         PG  2001     87.0
20       My Neighbor Totoro          G  1988     86.0
21        Napoleon Dynamite         PG  2004     96.0
22  Pee-wee's Big Adventure         PG  1985     91.0
23           Pretty in Pink      PG-13  1986     97.0
24       The Princess Bride         PG  1987     98.0
25                   Psycho          R  1960      NaN
26              Stand by Me          R  1986     89.0
27                  Super 8      PG-13  2011    112.0
28                 superbad          R  2007    113.0
29                 WarGames         PG  1983    114.0

In [ ]:
df.set_index('year',inplace=True)	#Change the index

task 

1- get any rows data by calling the index year 1986

In [ ]:
df

title     rating  runtime
year                                             
2000            Almost Famous          R    122.0
1999             American Pie          R     95.0
1985       Back to the Future         PG    116.0
1982             Blade Runner          R    117.0
1974        Blood for Dracula          R    106.0
1986              Blue Velvet          R    120.0
1985       The Breakfast Club          R     97.0
1995                 Clueless      PG-13      NaN
1967           Cool Hand Luke         GP    127.0
1996                The Craft          R    101.0
1965           Doctor Zhivago      PG-13    197.0
1970                  el Topo  Not Rated    125.0
1981                Evil Dead      NC-17     85.0
1984             Ghostbusters         PG    105.0
1978                   Grease      PG-13    110.0
1988                 Heathers          R    103.0
1986                Labyrinth         PG    101.0
1987            The Lost Boys          R     97.0
2004               Mean Girls      PG-13     97.0
2001       Millennium Actress         PG     87.0
1988       My Neighbor Totoro          G     86.0
2004        Napoleon Dynamite         PG     96.0
1985  Pee-wee's Big Adventure         PG     91.0
1986           Pretty in Pink      PG-13     97.0
1987       The Princess Bride         PG     98.0
1960                   Psycho          R      NaN
1986              Stand by Me          R     89.0
2011                  Super 8      PG-13    112.0
2007                 superbad          R    113.0
1983                 WarGames         PG    114.0

In [ ]:
df.loc[1986]

title rating  runtime
year                                
1986     Blue Velvet      R    120.0
1986       Labyrinth     PG    101.0
1986  Pretty in Pink  PG-13     97.0
1986     Stand by Me      R     89.0

In [ ]:
df[df.index == 1986]

title rating  runtime
year                                
1986     Blue Velvet      R    120.0
1986       Labyrinth     PG    101.0
1986  Pretty in Pink  PG-13     97.0
1986     Stand by Me      R     89.0

In [ ]:
df.reset_index(inplace=True)

In [ ]:
df

year                    title     rating  runtime
0   2000            Almost Famous          R    122.0
1   1999             American Pie          R     95.0
2   1985       Back to the Future         PG    116.0
3   1982             Blade Runner          R    117.0
4   1974        Blood for Dracula          R    106.0
5   1986              Blue Velvet          R    120.0
6   1985       The Breakfast Club          R     97.0
7   1995                 Clueless      PG-13      NaN
8   1967           Cool Hand Luke         GP    127.0
9   1996                The Craft          R    101.0
10  1965           Doctor Zhivago      PG-13    197.0
11  1970                  el Topo  Not Rated    125.0
12  1981                Evil Dead      NC-17     85.0
13  1984             Ghostbusters         PG    105.0
14  1978                   Grease      PG-13    110.0
15  1988                 Heathers          R    103.0
16  1986                Labyrinth         PG    101.0
17  1987            The Lost Boys          R     97.0
18  2004               Mean Girls      PG-13     97.0
19  2001       Millennium Actress         PG     87.0
20  1988       My Neighbor Totoro          G     86.0
21  2004        Napoleon Dynamite         PG     96.0
22  1985  Pee-wee's Big Adventure         PG     91.0
23  1986           Pretty in Pink      PG-13     97.0
24  1987       The Princess Bride         PG     98.0
25  1960                   Psycho          R      NaN
26  1986              Stand by Me          R     89.0
27  2011                  Super 8      PG-13    112.0
28  2007                 superbad          R    113.0
29  1983                 WarGames         PG    114.0

Task 



1- call movies with year more than 2000    



In [ ]:
df.set_index('year',inplace=True)

In [ ]:
df.loc[2000:] # Wrong ******

title     rating  runtime
year                                             
2000            Almost Famous          R    122.0
1999             American Pie          R     95.0
1985       Back to the Future         PG    116.0
1982             Blade Runner          R    117.0
1974        Blood for Dracula          R    106.0
1986              Blue Velvet          R    120.0
1985       The Breakfast Club          R     97.0
1995                 Clueless      PG-13      NaN
1967           Cool Hand Luke         GP    127.0
1996                The Craft          R    101.0
1965           Doctor Zhivago      PG-13    197.0
1970                  el Topo  Not Rated    125.0
1981                Evil Dead      NC-17     85.0
1984             Ghostbusters         PG    105.0
1978                   Grease      PG-13    110.0
1988                 Heathers          R    103.0
1986                Labyrinth         PG    101.0
1987            The Lost Boys          R     97.0
2004               Mean Girls      PG-13     97.0
2001       Millennium Actress         PG     87.0
1988       My Neighbor Totoro          G     86.0
2004        Napoleon Dynamite         PG     96.0
1985  Pee-wee's Big Adventure         PG     91.0
1986           Pretty in Pink      PG-13     97.0
1987       The Princess Bride         PG     98.0
1960                   Psycho          R      NaN
1986              Stand by Me          R     89.0
2011                  Super 8      PG-13    112.0
2007                 superbad          R    113.0
1983                 WarGames         PG    114.0

In [ ]:
df[df.index > 2000]

title rating  runtime
year                                    
2004          Mean Girls  PG-13     97.0
2001  Millennium Actress     PG     87.0
2004   Napoleon Dynamite     PG     96.0
2011             Super 8  PG-13    112.0
2007            superbad      R    113.0

In [ ]:
pd.set_option("display.max_rows", 50)


In [ ]:
ecom.head(75)

company_id purchase_date product_category  quantity  total_sales  gender  \
0            4    2021-12-19             Home         8   423.470537  Female   
1            3    2021-10-10           Beauty         2   394.399325  Female   
2            6    2021-06-28             Home         9   157.567488    Male   
3            2    2021-10-14            Books         1   113.351819    Male   
4            6    2020-02-09           Beauty         4    85.729982    Male   
..         ...           ...              ...       ...          ...     ...   
70           6    2021-08-23            Books         8   418.102069    Male   
71           1    2021-06-29           Beauty         1    75.348127    Male   
72           5    2021-02-11      Electronics         3   176.177484  Female   
73           2    2020-11-05            Books         5   490.617058  Female   
74           5    2021-11-29             Home         2    18.281618    Male   

   marital_status  price_per_unit       customer_type  age  
0        Divorced       19.412629        New Customer   54  
1        Divorced       34.296387        New Customer   71  
2         Widowed       15.469916        New Customer   80  
3         Married        8.158978  Returning Customer   26  
4          Single       26.598399        New Customer   46  
..            ...             ...                 ...  ...  
70         Single       27.964536  Returning Customer   33  
71         Single       22.160328  Returning Customer   54  
72       Divorced       32.177244        New Customer   73  
73        Widowed       43.603792  Returning Customer   46  
74        Widowed       14.666472  Returning Customer   54  

[75 rows x 10 columns]

In [ ]:
path = 'D:\Work\Cources\AI Academy\G4 Summer offline\Session 6/movies.csv'
data = pd.read_csv(path)
data.head()

title rating  year  runtime
0       Almost Famous      R  2000    122.0
1        American Pie      R  1999     95.0
2  Back to the Future     PG  1985    116.0
3        Blade Runner      R  1982    117.0
4   Blood for Dracula      R  1974    106.0

In [ ]:
df = data.copy(deep=True)

### 5- Filter, Sort, and Groupby



In [ ]:
# pd.set_option("display.max_rows", None)


In [ ]:
df.sample(2)


title rating  year  runtime
8      Cool Hand Luke     GP  1967    127.0
6  The Breakfast Club      R  1985     97.0

In [ ]:
df[df['year'] > 2000]	#Rows where the column col is greater than 2000

title rating  year  runtime
18          Mean Girls  PG-13  2004     97.0
19  Millennium Actress     PG  2001     87.0
21   Napoleon Dynamite     PG  2004     96.0
27             Super 8  PG-13  2011    112.0
28            superbad      R  2007    113.0

In [ ]:
df[df['runtime'] > 130]	#Rows where the column col is greater than 130

title rating  year  runtime
10  Doctor Zhivago  PG-13  1965    197.0

In [ ]:
df.isnull().sum()

title      0
rating     0
year       0
runtime    2
dtype: int64

df[  condtions queries   function return bool ]    == new df (after filter)

In [ ]:
df['rating'].value_counts()

rating
R            12
PG            8
PG-13         6
GP            1
Not Rated     1
NC-17         1
G             1
Name: count, dtype: int64

In [ ]:
df[df['runtime'].isnull()]

title rating  year  runtime
7   Clueless  PG-13  1995      NaN
25    Psycho      R  1960      NaN

In [ ]:
df[df['runtime'].isnull()]['rating'].value_counts()

rating
PG-13    1
R        1
Name: count, dtype: int64

In [ ]:
df[df['runtime'].notnull()]

& >>  And

| >>  or

~  >>  not

In [ ]:
df[(df['runtime'] > 95) & (df['runtime'] < 110)]	

title rating  year  runtime
4    Blood for Dracula      R  1974    106.0
6   The Breakfast Club      R  1985     97.0
9            The Craft      R  1996    101.0
13        Ghostbusters     PG  1984    105.0
15            Heathers      R  1988    103.0
16           Labyrinth     PG  1986    101.0
17       The Lost Boys      R  1987     97.0
18          Mean Girls  PG-13  2004     97.0
21   Napoleon Dynamite     PG  2004     96.0
23      Pretty in Pink  PG-13  1986     97.0
24  The Princess Bride     PG  1987     98.0

In [ ]:
# df[(df['runtime'] > 95) & ]	#Rows where 0.8 > col > 0.6

In [ ]:
df[(df['runtime'] > 95) &  (df['rating'] == "PG")]	

title rating  year  runtime
2   Back to the Future     PG  1985    116.0
13        Ghostbusters     PG  1984    105.0
16           Labyrinth     PG  1986    101.0
21   Napoleon Dynamite     PG  2004     96.0
24  The Princess Bride     PG  1987     98.0
29            WarGames     PG  1983    114.0

In [ ]:
# = # Assigning   

# == # are you equal  ,,,,,  Compare

Q 1

year after 2000  and runtime less than or equal 90 mins

In [ ]:
df[(df['year']>2000)&(df['runtime']<=90)]

title rating  year  runtime
19  Millennium Actress     PG  2001     87.0

task 



get movies  that runtime <90 or rating = R

In [ ]:
df.sort_values('year')	#Sort values by col1 in ascending order

title     rating  year  runtime
25                   Psycho          R  1960      NaN
10           Doctor Zhivago      PG-13  1965    197.0
8            Cool Hand Luke         GP  1967    127.0
11                  el Topo  Not Rated  1970    125.0
4         Blood for Dracula          R  1974    106.0
14                   Grease      PG-13  1978    110.0
12                Evil Dead      NC-17  1981     85.0
3              Blade Runner          R  1982    117.0
29                 WarGames         PG  1983    114.0
13             Ghostbusters         PG  1984    105.0
6        The Breakfast Club          R  1985     97.0
2        Back to the Future         PG  1985    116.0
22  Pee-wee's Big Adventure         PG  1985     91.0
5               Blue Velvet          R  1986    120.0
26              Stand by Me          R  1986     89.0
16                Labyrinth         PG  1986    101.0
23           Pretty in Pink      PG-13  1986     97.0
17            The Lost Boys          R  1987     97.0
24       The Princess Bride         PG  1987     98.0
15                 Heathers          R  1988    103.0
20       My Neighbor Totoro          G  1988     86.0
7                  Clueless      PG-13  1995      NaN
9                 The Craft          R  1996    101.0
1              American Pie          R  1999     95.0
0             Almost Famous          R  2000    122.0
19       Millennium Actress         PG  2001     87.0
18               Mean Girls      PG-13  2004     97.0
21        Napoleon Dynamite         PG  2004     96.0
28                 superbad          R  2007    113.0
27                  Super 8      PG-13  2011    112.0

In [ ]:
df.sort_values('year',ascending=False)	#Sort values by col2 in descending order.5

title     rating  year  runtime
27                  Super 8      PG-13  2011    112.0
28                 superbad          R  2007    113.0
21        Napoleon Dynamite         PG  2004     96.0
18               Mean Girls      PG-13  2004     97.0
19       Millennium Actress         PG  2001     87.0
0             Almost Famous          R  2000    122.0
1              American Pie          R  1999     95.0
9                 The Craft          R  1996    101.0
7                  Clueless      PG-13  1995      NaN
20       My Neighbor Totoro          G  1988     86.0
15                 Heathers          R  1988    103.0
17            The Lost Boys          R  1987     97.0
24       The Princess Bride         PG  1987     98.0
5               Blue Velvet          R  1986    120.0
26              Stand by Me          R  1986     89.0
16                Labyrinth         PG  1986    101.0
23           Pretty in Pink      PG-13  1986     97.0
22  Pee-wee's Big Adventure         PG  1985     91.0
2        Back to the Future         PG  1985    116.0
6        The Breakfast Club          R  1985     97.0
13             Ghostbusters         PG  1984    105.0
29                 WarGames         PG  1983    114.0
3              Blade Runner          R  1982    117.0
12                Evil Dead      NC-17  1981     85.0
14                   Grease      PG-13  1978    110.0
4         Blood for Dracula          R  1974    106.0
11                  el Topo  Not Rated  1970    125.0
8            Cool Hand Luke         GP  1967    127.0
10           Doctor Zhivago      PG-13  1965    197.0
25                   Psycho          R  1960      NaN

In [ ]:
df.sort_values(['year','rating'],
               ascending=[True,False])	#Sort values by col1 in ascending order then col2 in descending order

title     rating  year  runtime
25                   Psycho          R  1960      NaN
10           Doctor Zhivago      PG-13  1965    197.0
8            Cool Hand Luke         GP  1967    127.0
11                  el Topo  Not Rated  1970    125.0
4         Blood for Dracula          R  1974    106.0
14                   Grease      PG-13  1978    110.0
12                Evil Dead      NC-17  1981     85.0
3              Blade Runner          R  1982    117.0
29                 WarGames         PG  1983    114.0
13             Ghostbusters         PG  1984    105.0
6        The Breakfast Club          R  1985     97.0
2        Back to the Future         PG  1985    116.0
22  Pee-wee's Big Adventure         PG  1985     91.0
5               Blue Velvet          R  1986    120.0
26              Stand by Me          R  1986     89.0
23           Pretty in Pink      PG-13  1986     97.0
16                Labyrinth         PG  1986    101.0
17            The Lost Boys          R  1987     97.0
24       The Princess Bride         PG  1987     98.0
15                 Heathers          R  1988    103.0
20       My Neighbor Totoro          G  1988     86.0
7                  Clueless      PG-13  1995      NaN
9                 The Craft          R  1996    101.0
1              American Pie          R  1999     95.0
0             Almost Famous          R  2000    122.0
19       Millennium Actress         PG  2001     87.0
18               Mean Girls      PG-13  2004     97.0
21        Napoleon Dynamite         PG  2004     96.0
28                 superbad          R  2007    113.0
27                  Super 8      PG-13  2011    112.0

In [ ]:
df.groupby('year')['runtime'].mean()  #Aggregation

year
1960           NaN
1965    197.000000
1967    127.000000
1970    125.000000
1974    106.000000
1978    110.000000
1981     85.000000
1982    117.000000
1983    114.000000
1984    105.000000
1985    101.333333
1986    101.750000
1987     97.500000
1988     94.500000
1995           NaN
1996    101.000000
1999     95.000000
2000    122.000000
2001     87.000000
2004     96.500000
2007    113.000000
2011    112.000000
Name: runtime, dtype: float64

In [ ]:
df.groupby('rating')['runtime'].mean()

rating
G             86.000000
GP           127.000000
NC-17         85.000000
Not Rated    125.000000
PG           101.000000
PG-13        122.600000
R            105.454545
Name: runtime, dtype: float64

In [ ]:
df.groupby('rating')['runtime'].mean().idxmax()

'GP'

In [ ]:
df.groupby('rating')['runtime'].mean().max()

127.0

In [ ]:
df.groupby('rating')['runtime'].mean().idxmax(),df.groupby('rating')['runtime'].mean().max()

('GP', 127.0)

Q1 

how many are movies published per year and > 2002 ?

In [ ]:
df[df['year'] >= 2002].groupby('year')['title'].count()

year
2004    2
2007    1
2011    1
Name: title, dtype: int64

In [ ]:
df[df['year'] >= 2002].groupby('year')['title'].value_counts()

year  title            
2004  Mean Girls           1
      Napoleon Dynamite    1
2007  superbad             1
2011  Super 8              1
Name: count, dtype: int64

Task 

1- read ecommerce data

2- filter 

3- sort  by 2 columns

4- group by



In [ ]:
eccomrce = pd.read_csv('D:\Work\Cources\AI Academy\G4 Summer offline\Session 6/ecommerce_sales.csv')
eccomrce.sample(2)

Company_ID Purchase_Date Product_Category  Quantity  Total_Sales  \
7661           5    2021-07-08         Clothing         8   117.653529   
7778           5    2021-04-07            Books         6   301.174180   

      Gender Marital_Status  Price_Per_Unit Customer_Type  Age  
7661  Female        Widowed       15.568848  New Customer   43  
7778    Male       Divorced       18.829984  New Customer   42

In [ ]:
eccomrce.shape

(10000, 10)

Q1

print how many rows in data when quantity more than 9

In [ ]:
eccomrce[eccomrce['Quantity']> 9].shape[0]

952

Q2

get the average of total sales when quantity > 9 

In [ ]:
eccomrce[eccomrce['Quantity']> 9]['Total_Sales'].mean() #aggregation

248.3549436249295

sort by lowest to highest price 

In [ ]:
eccomrce.sort_values('Price_Per_Unit',ascending=True)

Company_ID Purchase_Date Product_Category  Quantity  Total_Sales  \
3673           6    2020-10-28             Home         6   103.023130   
4701           3    2021-10-29         Clothing         3   468.868008   
4117           1    2020-05-16            Books         2   311.691311   
1007           5    2021-05-23             Home         2   148.800696   
6543           5    2020-05-19      Electronics        10   370.217485   
...          ...           ...              ...       ...          ...   
4407           1    2020-08-28           Beauty         5   444.512669   
4341           2    2021-09-21           Beauty         4   460.393289   
9567           6    2020-04-18      Electronics        10   161.307900   
9425           5    2021-10-29         Clothing         4    34.949556   
5975           4    2020-06-10             Home         9    96.376067   

      Gender Marital_Status  Price_Per_Unit       Customer_Type  Age  
3673  Female         Single        5.005398  Returning Customer   28  
4701    Male        Married        5.007282  Returning Customer   40  
4117    Male       Divorced        5.008114        New Customer   74  
1007  Female        Married        5.009175  Returning Customer   45  
6543    Male        Widowed        5.012822        New Customer   65  
...      ...            ...             ...                 ...  ...  
4407    Male         Single       49.991285        New Customer   60  
4341    Male       Divorced       49.993228        New Customer   74  
9567  Female       Divorced       49.993439        New Customer   64  
9425  Female         Single       49.994285  Returning Customer   69  
5975  Female         Single       49.994841  Returning Customer   28  

[10000 rows x 10 columns]

get the average of total sales  for every product

In [ ]:
eccomrce['Total_Sales'].mean()

254.44772293800358

In [ ]:
eccomrce.groupby('Product_Category')['Total_Sales'].mean()

Product_Category
Beauty         254.507030
Books          254.932858
Clothing       252.714839
Electronics    252.799646
Home           257.278423
Name: Total_Sales, dtype: float64

get the average price for gender male

In [ ]:
eccomrce[eccomrce['Gender']=='Male']['Price_Per_Unit'].mean()

27.41393036480661

In [ ]:
eccomrce.groupby('Gender')['Price_Per_Unit'].mean().index[1] , eccomrce.groupby('Gender')['Price_Per_Unit'].mean()[1]

('Male', 27.41393036480661)

get the average price for gender female

In [ ]:
eccomrce[eccomrce['Gender']=='Female']['Price_Per_Unit'].mean()

27.5141134625341

get the sum of total sales when quantity > 5 and the customers are Divorced 

In [ ]:
eccomrce[(eccomrce['Quantity']>5) & (eccomrce['Marital_Status']=='Divorced')] ['Total_Sales'].sum()

304195.759684742

In [ ]:
df.pivot_table(index='year',  # Categorical
               columns=['rating'], # Categorical
               values=['runtime'], # numerical
               aggfunc='count' ,
               fill_value=0
               )	

#Create a pivot table that groups by col1 and calculates the mean of col2 and col3

runtime                               
rating       G GP NC-17 Not Rated PG PG-13  R
year                                         
1960         0  0     0         0  0     0  0
1965         0  0     0         0  0     1  0
1967         0  1     0         0  0     0  0
1970         0  0     0         1  0     0  0
1974         0  0     0         0  0     0  1
1978         0  0     0         0  0     1  0
1981         0  0     1         0  0     0  0
1982         0  0     0         0  0     0  1
1983         0  0     0         0  1     0  0
1984         0  0     0         0  1     0  0
1985         0  0     0         0  2     0  1
1986         0  0     0         0  1     1  2
1987         0  0     0         0  1     0  1
1988         1  0     0         0  0     0  1
1995         0  0     0         0  0     0  0
1996         0  0     0         0  0     0  1
1999         0  0     0         0  0     0  1
2000         0  0     0         0  0     0  1
2001         0  0     0         0  1     0  0
2004         0  0     0         0  1     1  0
2007         0  0     0         0  0     0  1
2011         0  0     0         0  0     1  0

In [ ]:
df1 = df.pivot_table(
               index='year',  # Categorical
               columns=['rating'], # Categorical
               values=['runtime'], # numerical
               aggfunc='count' 
               )
df1['total'] = df1.sum(axis=1)
df1

runtime                                      total
rating       G   GP NC-17 Not Rated   PG PG-13    R      
year                                                     
1960       NaN  NaN   NaN       NaN  NaN   NaN  0.0   0.0
1965       NaN  NaN   NaN       NaN  NaN   1.0  NaN   1.0
1967       NaN  1.0   NaN       NaN  NaN   NaN  NaN   1.0
1970       NaN  NaN   NaN       1.0  NaN   NaN  NaN   1.0
1974       NaN  NaN   NaN       NaN  NaN   NaN  1.0   1.0
1978       NaN  NaN   NaN       NaN  NaN   1.0  NaN   1.0
1981       NaN  NaN   1.0       NaN  NaN   NaN  NaN   1.0
1982       NaN  NaN   NaN       NaN  NaN   NaN  1.0   1.0
1983       NaN  NaN   NaN       NaN  1.0   NaN  NaN   1.0
1984       NaN  NaN   NaN       NaN  1.0   NaN  NaN   1.0
1985       NaN  NaN   NaN       NaN  2.0   NaN  1.0   3.0
1986       NaN  NaN   NaN       NaN  1.0   1.0  2.0   4.0
1987       NaN  NaN   NaN       NaN  1.0   NaN  1.0   2.0
1988       1.0  NaN   NaN       NaN  NaN   NaN  1.0   2.0
1995       NaN  NaN   NaN       NaN  NaN   0.0  NaN   0.0
1996       NaN  NaN   NaN       NaN  NaN   NaN  1.0   1.0
1999       NaN  NaN   NaN       NaN  NaN   NaN  1.0   1.0
2000       NaN  NaN   NaN       NaN  NaN   NaN  1.0   1.0
2001       NaN  NaN   NaN       NaN  1.0   NaN  NaN   1.0
2004       NaN  NaN   NaN       NaN  1.0   1.0  NaN   2.0
2007       NaN  NaN   NaN       NaN  NaN   NaN  1.0   1.0
2011       NaN  NaN   NaN       NaN  NaN   1.0  NaN   1.0

In [ ]:
# df.groupby('rating')['runtime'].count()

rating
G             1
GP            1
NC-17         1
Not Rated     1
PG            8
PG-13         5
R            11
Name: runtime, dtype: int64

Task 

try pivot table on ecomerce data with mean func

In [ ]:
ecom.columns

Index(['company_id', 'purchase_date', 'product_category', 'quantity',
       'total_sales', 'gender', 'marital_status', 'price_per_unit',
       'customer_type', 'age'],
      dtype='object')

In [ ]:
ecom.pivot_table(index=['product_category','customer_type'],  # Categorical
                columns=['gender','marital_status'], # Categorical
                values='total_sales', # numerical
                aggfunc='mean' ,
                fill_value=0
                )	

#Create a pivot table that groups by col1 and calculates the mean of col2 and col3

gender                                   Female                          \
marital_status                         Divorced     Married      Single   
product_category customer_type                                            
Beauty           New Customer        244.910269  257.508473  245.397777   
                 Returning Customer  251.961225  259.604287  263.232640   
Books            New Customer        246.168904  255.454812  246.827849   
                 Returning Customer  262.441560  273.260704  255.723058   
Clothing         New Customer        277.175094  269.096805  244.433178   
                 Returning Customer  278.054728  224.198744  235.142284   
Electronics      New Customer        243.523048  251.396153  247.690528   
                 Returning Customer  255.817035  264.391669  249.495194   
Home             New Customer        231.724022  274.577982  263.423205   
                 Returning Customer  262.908570  241.009977  241.028546   

gender                                                 Male              \
marital_status                          Widowed    Divorced     Married   
product_category customer_type                                            
Beauty           New Customer        260.760709  245.008594  256.701519   
                 Returning Customer  263.324738  262.630434  251.803349   
Books            New Customer        257.410864  269.504352  248.009820   
                 Returning Customer  275.696483  248.429466  240.157092   
Clothing         New Customer        251.018495  239.360797  253.835825   
                 Returning Customer  248.039524  255.044586  270.431620   
Electronics      New Customer        253.713424  242.814490  234.363178   
                 Returning Customer  233.823685  244.490442  267.675146   
Home             New Customer        274.811015  260.774112  254.063335   
                 Returning Customer  251.947635  261.882473  239.303682   

gender                                                       
marital_status                           Single     Widowed  
product_category customer_type                               
Beauty           New Customer        274.654131  251.407427  
                 Returning Customer  257.504992  232.725523  
Books            New Customer        263.173715  235.308673  
                 Returning Customer  249.036792  254.196346  
Clothing         New Customer        260.829996  239.959875  
                 Returning Customer  262.795140  241.411263  
Electronics      New Customer        273.090521  263.928573  
                 Returning Customer  252.617732  265.383541  
Home             New Customer        263.254182  269.470837  
                 Returning Customer  266.666068  257.988853

In [ ]:
ecom.corr(numeric_only=True)

company_id  quantity  total_sales  price_per_unit       age
company_id        1.000000 -0.017297     0.012920       -0.008498 -0.016260
quantity         -0.017297  1.000000    -0.011340        0.006762  0.013805
total_sales       0.012920 -0.011340     1.000000       -0.008614  0.002354
price_per_unit   -0.008498  0.006762    -0.008614        1.000000  0.005603
age              -0.016260  0.013805     0.002354        0.005603  1.000000

# Task:

- Read this data 
- Print some analysis 
- Apply one pivot table

### 6- Join/Concat



![join.png](attachment:join.png)

In [ ]:
import pandas as pd

In [ ]:
row_A1 = {'Nid':1,'Add':'Alex','Age':25}
row_A2 = {'Nid':2,'Add':'Cairo','Age':30}
row_A3 = {'Nid':4,'Add':'behira','Age':55}
row_A4 = {'Nid':10,'Add':'menof','Age':64}
row_A5 = {'Nid':15,'Add':'alamin','Age':20}
row_A6 = {'Nid':20,'Add':'Aswan','Age':15}

Data_A = pd.DataFrame([row_A1,row_A2,row_A3,
                        row_A4,row_A5,row_A6])

Data_A

Nid     Add  Age
0    1    Alex   25
1    2   Cairo   30
2    4  behira   55
3   10   menof   64
4   15  alamin   20
5   20   Aswan   15

In [ ]:
row_B1 = {'Nid':20,'Sal':5000,'Mother':'Sara'}
row_B2 = {'Nid':10,'Sal':4500,'Mother':'Eman'}
row_B3 = {'Nid':16,'Sal':3000,'Mother':'Aml'}
row_B4 = {'Nid':5,'Sal':10000,'Mother':"Rania"}
row_B5 = {'Nid':3,'Sal':9700,'Mother':'Laila'}
row_B6 = {'Nid':2,'Sal':20000,'Mother':'Sahar'}
row_B7 = {'Nid':4,'Sal':6500,'Mother':'Hend'}

Data_B = pd.DataFrame([row_B1,row_B2,row_B3,
                        row_B4,row_B5,row_B6,
                        row_B7])

Data_B

Nid    Sal Mother
0   20   5000   Sara
1   10   4500   Eman
2   16   3000    Aml
3    5  10000  Rania
4    3   9700  Laila
5    2  20000  Sahar
6    4   6500   Hend

In [ ]:
Data_A

Nid     Add  Age
0    1    Alex   25
1    2   Cairo   30
2    4  behira   55
3   10   menof   64
4   15  alamin   20
5   20   Aswan   15

In [ ]:
Data_A.merge(Data_B,
            how='inner',
            on = 'Nid',
            # left_on='Nid',
            # right_on='Nid',
            )

Nid     Add  Age    Sal Mother
0    2   Cairo   30  20000  Sahar
1    4  behira   55   6500   Hend
2   10   menof   64   4500   Eman
3   20   Aswan   15   5000   Sara

In [ ]:
Data_A.merge(Data_B,
            how='left',
            on='Nid'
            )

Nid     Add  Age      Sal Mother
0    1    Alex   25      NaN    NaN
1    2   Cairo   30  20000.0  Sahar
2    4  behira   55   6500.0   Hend
3   10   menof   64   4500.0   Eman
4   15  alamin   20      NaN    NaN
5   20   Aswan   15   5000.0   Sara

In [ ]:
Data_A.merge(Data_B,
            how='right',
            on='Nid',
            copy=False)

Nid     Add   Age    Sal Mother
0   20   Aswan  15.0   5000   Sara
1   10   menof  64.0   4500   Eman
2   16     NaN   NaN   3000    Aml
3    5     NaN   NaN  10000  Rania
4    3     NaN   NaN   9700  Laila
5    2   Cairo  30.0  20000  Sahar
6    4  behira  55.0   6500   Hend

In [ ]:
Data_A.merge(Data_B,
            how='outer',
            on='Nid',
                )

Nid     Add   Age      Sal Mother
0    1    Alex  25.0      NaN    NaN
1    2   Cairo  30.0  20000.0  Sahar
2    4  behira  55.0   6500.0   Hend
3   10   menof  64.0   4500.0   Eman
4   15  alamin  20.0      NaN    NaN
5   20   Aswan  15.0   5000.0   Sara
6   16     NaN   NaN   3000.0    Aml
7    5     NaN   NaN  10000.0  Rania
8    3     NaN   NaN   9700.0  Laila

In [ ]:
Data_A

Nid     Add  Age
0    1    Alex   25
1    2   Cairo   30
2    4  behira   55
3   10   menof   64
4   15  alamin   20
5   20   Aswan   15

In [ ]:
Data_A1 = Data_A.iloc[:-1]
Data_A1

Nid     Add  Age
0    1    Alex   25
1    2   Cairo   30
2    4  behira   55
3   10   menof   64
4   15  alamin   20

![concat.png](attachment:concat.png)

In [ ]:
pd.concat([Data_A, Data_A1],
            axis=0)	# axis 1    besides  ,  axis 0  over

#Add the columns in df1 to the end of df2 (columns should be identical)

Nid     Add  Age
0    1    Alex   25
1    2   Cairo   30
2    4  behira   55
3   10   menof   64
4   15  alamin   20
5   20   Aswan   15
0    1    Alex   25
1    2   Cairo   30
2    4  behira   55
3   10   menof   64
4   15  alamin   20

In [ ]:
row_A1 = {'Nid':1,'Add':'Alex','Age':25}
row_A2 = {'Nid':2,'Add':'Cairo','Age':30}
row_A3 = {'Nid':4,'Add':'behira','Age':55}


Data_A1 = pd.DataFrame([row_A1,row_A2,row_A3,
                        ])

Data_A1

Nid     Add  Age
0    1    Alex   25
1    2   Cairo   30
2    4  behira   55

In [ ]:
row_A4 = {'Nid':10,'Add':'menof','Age':64}
row_A5 = {'Nid':15,'Add':'alamin','Age':20}
row_A6 = {'Nid':20,'Add':'Aswan','Age':15}


Data_A2 = pd.DataFrame([row_A4,row_A5,row_A6,
                        ])

Data_A2

Nid     Add  Age
0   10   menof   64
1   15  alamin   20
2   20   Aswan   15

In [ ]:
pd.concat([Data_A1, Data_A2],
            axis=0,
            ignore_index=True
            )	

Nid     Add  Age
0    1    Alex   25
1    2   Cairo   30
2    4  behira   55
3   10   menof   64
4   15  alamin   20
5   20   Aswan   15

In [ ]:
Data_A1.merge(Data_A2,how='outer')

Nid     Add  Age
0    1    Alex   25
1    2   Cairo   30
2    4  behira   55
3   10   menof   64
4   15  alamin   20
5   20   Aswan   15

### 7- Exporting Data

In [ ]:
df_merge = Data_A.merge(Data_B,
                        how='inner',
                        on='Nid')

In [ ]:
df_merge

Nid     Add  Age    Sal Mother
0    2   Cairo   30  20000  Sahar
1    4  behira   55   6500   Hend
2   10   menof   64   4500   Eman
3   20   Aswan   15   5000   Sara

In [ ]:
df_merge.to_csv('merge.csv')	#Write to a CSV file

Task 

export to excel

In [ ]:
# df.to_excel(filename.xlsx) #	Write to an Excel file

Task 

export to pickle

and read

In [ ]:
df_merge.to_pickle('merge.pkl')

In [ ]:
merg_pkl = pd.read_pickle('merge.pkl')
merg_pkl

Nid     Add  Age    Sal Mother
0    2   Cairo   30  20000  Sahar
1    4  behira   55   6500   Hend
2   10   menof   64   4500   Eman
3   20   Aswan   15   5000   Sara

# Visualization 

# Big Data Anaylsis  > SQL   and Linux   on VM 
